modules/proxy/mod_proxy.h

```diff
@@ -467,7 +467,7 @@ typedef struct {
     apr_size_t      recv_buffer_size;
     apr_size_t      io_buffer_size;
     apr_size_t      elected;    /* Number of times the worker was elected */
-    apr_size_t      busy;       /* busyness factor */
+    volatile apr_size_t      busy;       /* busyness factor */
```

Comment:
It's public interface so I think that it can't change without a MMN major bump. But I don't think `volatile` is actually needed here if worker->s->busy is **always** accessed with the new atomic functions (i.e. we shouldn't use bare worker->s->busy anywhere). Looks like one case is missing in proxy_balancer_get_best_worker(), not sure if the other accesses to worker->s->busy in the other lbmethods (!bybusiness) should use the new helpers for consistency too, looks sensible if the common proxy_balancer_get_best_worker() has to.

Reply:
Since the function parameter is declared volatile, any passed in memory (whether or not declared volatile itself) will be considered volatile at the function call and within the function itself (i.e. the volatile qualifier is additive, not part of the type checking). But volatility is not enough (per se) to provide atomicity on all platforms, that's why we can't use worker->s->busy directly and need the *_busy_count() helpers to read/change it. So we'll never use the volatile qualifier on proxy_worker_shared::busy anyway if it existed, IOW if the **accesses** to worker->s->busy are all made volatile then the memory itself needs not be declared volatile.
And I think that adding the volatile qualifier to some public struct member is an API change which requires an MMN **major** bump (some third party code using it might not compile anymore because the compiler could then warn about existing code "removing" the volatile qualification somehow), meaning it's not backportable to 2.4.x. That's why I'm trying to avoid it..

Reply:
fixed